### PR TITLE
feat(metadata): Allow querying by metadata.

### PR DESCRIPTION
**Pull request by christiandompierre**
Adds the possibility to also filter by metadata in `get_with_params`
Added `list_with_params` to let you filter when you list.

Relates to https://github.com/uc-cdis/indexd/pull/85

### CODE REVIEW

File: indexclient/client.py [hunk 1]
```diff
@@ -85,22 +85,24 @@ def get(self, did):
 
         return Document(self, did, json=response.json())
 
-    def get_with_params(self, params=None):
+    def get_with_params(self, params={}):
```
Comment (philloooo) [inline indexclient/client.py]: using {} for default parameter should be avoided

File: indexclient/client.py [hunk 2]
```diff
@@ -116,10 +118,26 @@ def get_with_params(self, params=None):
 
     def list(self, limit=float("inf"), start=None, page_size=100):
         """ Returns a generator of document objects. """
+        return self.list_with_params(limit, start, page_size)
+
+    def list_with_params(self, limit=float("inf"), start=None, page_size=100, params={}):
```
Comment (philloooo) [inline indexclient/client.py]: unit tests

File: indexclient/client.py [hunk 3]
```diff
@@ -88,19 +88,23 @@ def get(self, did):
...
+            params_copy['hash'] = params_copy.pop('hashes')
+        reformatted_params = dict()
+        for param in ['hash', 'metadata']:
+            if param in params_copy:
+                reformatted_params[param] = []
+                for k, v in params_copy[param].items():
+                    reformatted_params[param].append(str(k) + ':' + str(v))
+                del params_copy[param]
         reformatted_params.update(params_copy)
+        reformatted_params['limit'] = 1
```
Comment (jbarno) [inline indexclient/client.py]: Shouldn't this come from the params? Or indicate that get_with_params will only ever return 1?

**Comment by christiandompierre** [comment]
I am not changing the behavior, it's already only returning 1 result.

File: tests/conftest.py [hunk 4]
```diff
@@ -1,10 +1,7 @@
 import pytest
 
-from cdisutilstest.code.indexd_fixture import indexd_server
-from indexclient.client import IndexClient
+from cdisutilstest.code.conftest import *
```
Comment (philloooo) [inline tests/conftest.py]: avoid importing *

**Pull request merged**